src/xenia/cpu/xex_module.h

```diff
@@ -30,61 +29,158 @@ class Runtime;
+
+  uint32_t image_size() const {
+    assert_not_zero(base_address_);
+
+    // Calculate the new total size of the XEX image from its headers.
+    auto heap = memory()->LookupHeap(base_address_);
+    uint32_t total_size = 0;
+    for (uint32_t i = 0; i < xex_security_info()->page_descriptor_count; i++) {
+      // Byteswap the bitfield manually.
+      xex2_page_descriptor desc;
+      desc.value =
+          xe::byte_swap(xex_security_info()->page_descriptors[i].value);
+
+      total_size += desc.size * heap->page_size();
+    }
+    return total_size;
+  }
+
+  const uint8_t* session_key() const { return session_key_; }
+
+  const std::vector<ImportLibrary>* import_libraries() const {
+    return &import_libs_;
+  }
+
+  const xex2_opt_execution_info* opt_execution_info() const {
+    xex2_opt_execution_info* retval = nullptr;
+    GetOptHeader(XEX_HEADER_EXECUTION_INFO, &retval);
+    return retval;
+  }
+
+  const xex2_opt_file_format_info* opt_file_format_info() const {
+    xex2_opt_file_format_info* retval = nullptr;
+    GetOptHeader(XEX_HEADER_FILE_FORMAT_INFO, &retval);
+    return retval;
+  }
+
   // Gets an optional header. Returns NULL if not found.
   // Special case: if key & 0xFF == 0x00, this function will return the value,
   // not a pointer! This assumes out_ptr points to uint32_t.
-  static bool GetOptHeader(const xex2_header* header, xe_xex2_header_keys key,
+  static bool GetOptHeader(const xex2_header* header, xex2_header_keys key,
                            void** out_ptr);
-  bool GetOptHeader(xe_xex2_header_keys key, void** out_ptr) const;
+  bool GetOptHeader(xex2_header_keys key, void** out_ptr) const;
 
   // Ultra-cool templated version
   // Special case: if key & 0xFF == 0x00, this function will return the value,
   // not a pointer! This assumes out_ptr points to uint32_t.
   template <typename T>
-  static bool GetOptHeader(const xex2_header* header, xe_xex2_header_keys key,
+  static bool GetOptHeader(const xex2_header* header, xex2_header_keys key,
                            T* out_ptr) {
     return GetOptHeader(header, key, reinterpret_cast<void**>(out_ptr));
   }
 
   template <typename T>
-  bool GetOptHeader(xe_xex2_header_keys key, T* out_ptr) const {
+  bool GetOptHeader(xex2_header_keys key, T* out_ptr) const {
     return GetOptHeader(key, reinterpret_cast<void**>(out_ptr));
   }
 
   static const xex2_security_info* GetSecurityInfo(const xex2_header* header);
 
+  const PESection* GetPESection(const char* name);
+
   uint32_t GetProcAddress(uint16_t ordinal) const;
   uint32_t GetProcAddress(const char* name) const;
 
-  bool ApplyPatch(XexModule* module);
   bool Load(const std::string& name, const std::string& path,
-            const void* xex_addr, size_t xex_length);
-  bool Load(const std::string& name, const std::string& path, xe_xex2_ref xex);
+            const void* xex_addr, size_t xex_length,
+            XexModule* base_xex = nullptr);
```

Comment:
What made you decide to load patches on top of their base XEX files?
It looks like you're doing this to reuse a lot of the loader code, and I'd like to hear your thoughts. I originally wanted to have an ApplyPatch function.

Reply:
Well I did start by adding to ApplyPatch, but when I noticed it was doing most of what Load/ReadImage already did I decided an overlay approach would be better (also at that time though I thought the ReadImageCompressed function could be re-used with some small tweaks, which turned out not to be the case...)

Changing it to back to be like ApplyPatch might actually be better, wouldn't have to keep hold of base_xex_ or anything. I'll try looking into it now.